tests/integration/framework/docker_runner.py

```diff
@@ -8,49 +8,93 @@
         """Sets platform and image for all tests ran with this instance"""
         self.platform = platform
         self.image = image
         self.verbose = verbose
+        self.container_id = None
 
-    def construct_docker_command(self, envs, args):
+        # Launch the container
+        self._start()
+
+    def execute(self, envs, args):
+        """Launch `docker exec` commands inside the background container"""
+        command = self._construct_command("exec", envs, args)
+        return self._shell(command)
+
+    def run(self, envs, args):
+        """
+        Launch `docker run` commands to create a new container for each command
+        """
+        command = self._construct_command("run", envs, args)
+        return self._shell(command)
+
+    def __del__(self):
+        """Remove test container"""
+        stop_command = ["docker", "rm", "-f", self.container_id]
+        self._shell(stop_command, silent=True)
+
+    def _start(self):
         """
-        Construct a docker command with env and args
+        Launch a docker container. Done in 2 step using create + start.
+
+        Keep the container running with a shell open thanks to `--interactive`,
+        having stdin open keep the process running.
         """
-        command = ["docker", "run", "--platform", self.platform]
+        create_command = [
+                "docker", "create",
+                "--platform", self.platform,
+                "--interactive",
+                self.image,
+                "/bin/bash",
+                ]
+        self.container_id = self._shell(create_command, silent=True)
+
+        start_command = [
+                "docker", "start", self.container_id
+                ]
+        self._shell(start_command, silent=True)
+
+    def _shell(self, command, silent=False):
+        """Run an arbitrary shell command and return its output"""
+        if self.verbose and not silent:
+            print(f"$ { ' '.join(command) }")
+
+        try:
+            result = subprocess.run(command, capture_output=True, check=True)
+        except subprocess.CalledProcessError as command_err:
+            print(command_err.stdout.decode("utf-8"), file=sys.stdout)
+            print(command_err.stderr.decode("utf-8"), file=sys.stderr)
+            raise command_err
+
+        return result.stdout.decode("utf-8").strip()
+
+    def _construct_command(self, docker_cmd, envs, args):
+        """Construct a docker command with env and args"""
+        command = ["docker", docker_cmd]
 
         for env in envs:
             command.append("-e")
             command.append(env)
 
-        command.append(self.image)
-
-        for arg in args:
-            command.append(arg)
-
-        return command
+        # Use container or image depending on command type
+        if docker_cmd == "exec":
+            tag = self.container_id
+        elif docker_cmd == "run":
+            tag = self.image
 
-    def run_interactive_command(self, envs, args):
```

Comment:
could you please bring this method back so that the tests don't have to change?

Reply:
Related to https://github.com/dogecoin/docker/pull/51#discussion_r771949065